### PR TITLE
Add climate reproducibility test for MPAS Ocean

### PR DESCRIPTION
**Pull request by mkstratos**
Add a multi-variate Kolmogorov–Smirnov test for the MPAS Ocean component.

Based on experiments conducted for Mahajan (2021) doi:10.1145/3468267.3470572.
Because the test is specific to the MPAS-Ocean component, it's placed in the `components/mpas-ocean/cime_config/SystemTests` directory. This is initialized with a restart file from a 400+ year run with the `GMPAS-NYF` compset.

### CODE REVIEW

**Comment by mkstratos** [comment]
Example passing test: https://web.lcrc.anl.gov/public/e3sm/ac.mkelleher/evv/MVKO_PS.T62_oQU240.GMPAS-NYF.chrysalis_intel.C.20240131_1022/
Example failing test: https://web.lcrc.anl.gov/public/e3sm/ac.mkelleher/evv/MVKO_PS.T62_oQU240.GMPAS-NYF.chrysalis_intel.C.20240131_0945/

File: cime_config/config_files.xml [hunk 1]
```diff
@@ -98,6 +98,7 @@
     <type>char</type>
     <values>
       <value>$CIMEROOT/CIME/data/config/config_tests.xml</value>
+      <value component="mpaso">$COMP_ROOT_DIR_OCN/cime_config/config_tests.xml</value>
```
[the remark this answers is not in the document]
Comment (jasonb5) [reply cime_config/config_files.xml:101]: Yes I believe this is the only to add a component specific test.

**Comment by rljacob** [comment]
@mkstratos how do you run this test?

**Comment by mkstratos** [comment]
The initial condition file is currently only on LCRC, so from either Chrysalis or Anvil it can be run on this fork/branch with:
`${E3SM_ROOT}/cime/scripts/create_test MVKO_PS.T62_oQU240.GMPAS-NYF [-g / -c] --pesfile ${E3SM_ROOT}/cime_config/testmods_dirs/config_pes_tests.xml`

**Comment by rljacob** [comment]
Next question, if you create a suite in tests.py called e3sm_ocn_nbfb and add this test, can "create_test e3sm_ocn_nbfb"  execute the test?

**Comment by mkstratos** [comment]
Yes, that would work

**Comment by rljacob** [comment]
Are you sure?  Please try it and let me know and I'll approve this PR.

**Comment by mkstratos** [comment]
I've confirmed this works, should we add that that suite to this PR?

**Comment by rljacob** [comment]
Yes please do.  Also please add a new "e3sm_nbfb" suite that combines that one and the atm one.

**Comment by rljacob** [comment]
What does the full test name that you tried look like?

**Review by rljacob — APPROVED** [review]

File: cime_config/config_files.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (rljacob) [reply cime_config/config_files.xml:101]: I also wasn't sure our python-dictionary based test-suite definition (tests.py) would find that test but it did.

**Comment by mark-petersen** [comment]
I am not familiar with this testing method, so removed myself as a reviewer.

**Comment by xylar** [comment]
@mkstratos, could you let me know what would be some helpful testing for me to do on this PR?

Also, it seems the test uses the oQU240 mesh, which is too coarse to be scientifically valid.  Could you let us know if that is an issue or if you are confident that the climate reproducibility (or lack there of) at such a coarse resolution would translate to a scientifically valid resolution as well?

**Comment by xylar** [comment]
@mkstratos, I finally found some time to test this.  Unfortunately, it immediately failed for me (probably user error) on Chrysalis with:
```
$ cd cime/scripts
$ ./create_test MVKO_PS.T62_oQU240.GMPAS-NYF -g /lcrc/group/e3sm/ac.xylar/e3sm_baselines/test_20240404 --pesfile ../../cime_config/testmods_dirs/config_pes_tests.xml --wait

Testnames: ['MVKO_PS.T62_oQU240.GMPAS-NYF.chrysalis_intel']
Using project from config_machines.xml: e3sm
create_test will do up to 1 tasks simultaneously
create_test will use up to 160 cores simultaneously
Creating test directory /lcrc/group/e3sm/ac.xasay-davis/scratch/chrys/MVKO_PS.T62_oQU240.GMPAS-NYF.chrysalis_intel.G.20240404_102105_tts9dt
RUNNING TESTS:
  MVKO_PS.T62_oQU240.GMPAS-NYF.chrysalis_intel
Starting CREATE_NEWCASE for test MVKO_PS.T62_oQU240.GMPAS-NYF.chrysalis_intel with 1 procs
Finished CREATE_NEWCASE for test MVKO_PS.T62_oQU240.GMPAS-NYF.chrysalis_intel in 0.175126 seconds (FAIL). [COMPLETED 1 of 1]
    Case dir: /lcrc/group/e3sm/ac.xasay-davis/scratch/chrys/MVKO_PS.T62_oQU240.GMPAS-NYF.chrysalis_intel.G.20240404_102105_tts9dt
    Errors were:
        ERROR: Makes no sense to have empty read-only file: /gpfs/fs1/home/ac.xylar/e3sm_work/E3SM/mkstratos/tests/add-mpaso-mvk-test/driver-nuopc/cime_config/config_component.xml

Waiting for tests to finish
FAIL MVKO_PS.T62_oQU240.GMPAS-NYF.chrysalis_intel (phase CREATE_NEWCASE)
    Case dir: /lcrc/group/e3sm/ac.xasay-davis/scratch/chrys/MVKO_PS.T62_oQU240.GMPAS-NYF.chrysalis_intel.G.20240404_102105_tts9dt
test-scheduler took 1.212627649307251 seconds
```
I tried merging `master` in case that made a difference but it didn't.  Happy to continue testing if you can give me some hints about what I'm doing wrong.

**Comment by xylar** [comment]
Maybe some updates are needed since the MOAB driver was added?

**Comment by rljacob** [comment]
It should use the MCT driver and nothing in the MOAB one will matter.  This string in the error message "driver-nuopc" is suspicious.   The test thinks its using the nuopc driver which we don't have.   I wonder if the "V" in "MVKO" is confusing create_test because "V" is a testtype mod CIME uses to change the driver.  But you would have to name the test "MVKO_Vnuopc_PS.T62_oQU240.GMPAS-NYF" to get that.

**Comment by xylar** [comment]
@rljacob, this isn't the first time I've seen this kind of error but I don't remember what I did before to get it.  Maybe leaving off the machine and compiler?

**Comment by rljacob** [comment]
Do you have CIME_MODEL env variable set to something besides "e3sm"  ?

**Comment by xylar** [comment]
No and I explicitly set it to e3sm, which also made no difference. 

Also, I can run other tests just fine.

**Comment by xylar** [comment]
No and I explicitly set it to e3sm, which also made no difference. 

Also, I can run other tests just fine.

**Comment by rljacob** [comment]
@xylar did you update submodules after switching to the branch?   I got past create_newcase.  Now have a python error which I think is because I'm not in the cime_env environment.

**Comment by rljacob** [comment]
Test built and submitted but asked for 120 nodes even with PS?  Is that right?
```
             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
            497642   compute test.MVK    jacob PD       0:00    120 (Resources)
```

**Comment by xylar** [comment]
@rljacob, oh, I didn't realize I needed to `source /lcrc/soft/climate/e3sm-unified/load_latest_cime_env.sh` to run this.  I guess that makes sense.

**Comment by xylar** [comment]
@rljacob
> did you update submodules after switching to the branch?

Yes, I did that.  I just did a hard reset back to this branch as it is, and updated the submodules again.  I ran:
```
source /lcrc/soft/climate/e3sm-unified/load_latest_cime_env.sh
./create_test MVKO_PS.T62_oQU240.GMPAS-NYF.chrysalis_intel -g /lcrc/group/e3sm/ac.xylar/e3sm_baselines/test_20240404 --wait
```
Same results as above...

**Comment by xylar** [comment]
I'm going to try to start fresh.  Maybe something got messed up along the way with my worktree.

**Comment by xylar** [comment]
I know what it is!  This happened to me before!  You can't work from a directory that has `tests` in the name, it confuses CIME.

@mkstratos, please use something other that `tests` in your branch names in the future, as tempting as it is...

**Comment by xylar** [comment]
Yep, it's working fine now that I took `tests` out of my directory structure.

**Comment by jgfouca** [comment]
@xylar , I made a CIME issue: https://github.com/ESMCI/cime/issues/4611

**Comment by xylar** [comment]
Thanks for doing that @jgfouca!

**Comment by xylar** [comment]
@mkstratos and @rljacob, I think something is still not quite right with the configuration.

First, I ran without the `--pesfile` flag (pompously disregarding your instructions above).  This led to trying to run the test on 120 nodes!

Then, I learned my lesson and ran:
```
$ source /lcrc/soft/climate/e3sm-unified/load_latest_cime_env.sh
$ ./create_test MVKO_PS.T62_oQU240.GMPAS-NYF.chrysalis_intel -g \
   -b /lcrc/group/e3sm/ac.xylar/e3sm_baselines/test_20240404 --wait -o \
   --pesfile ../../cime_config/testmods_dirs/config_pes_tests.xml
```
But the test is still running on 30 nodes!
```
$ squeue -u $USER
             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON) 
            497653   compute test.MVK ac.xasay PD       0:00     30 (Priority)
```
I don't think that can be right.  Any idea what's causing this?

**Comment by xylar** [comment]
Oh, sorry, I see.  It's running 30 runs each on 32 cores all at once.  Sorry, I missed that part of the design.

File: components/mpas-ocean/cime_config/SystemTests/mvko.py [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+
+        init_file_path = os.path.join(
+            self._case.get_value("DIN_LOC_ROOT"),
+            "ocn/mpas-o/oQU240/inic/gmpas-nyf",
+            INIT_FILE_NAME,
+        )
+        # Write yearly averages to custom output file
+        tss_climatology_config = [
+            "config_am_timeseriesstatsclimatology_enable = .true.\n",
+            "config_am_timeseriesstatsclimatology_backward_output_offset = '00-03-00_00:00:00'\n",
```
[the remark this answers is not in the document]
Comment (xylar) [reply components/mpas-ocean/cime_config/SystemTests/mvko.py:183]: Once this is fixed, I'll approve.

**Comment by xylar** [comment]
I compared this branch with a merge into `master`.  The two were statistically identical (great news!):
https://web.lcrc.anl.gov/public/e3sm/ac.xasay-davis/evv/MVKO_PS.T62_oQU240.GMPAS-NYF.chrysalis_intel.C.20240404_151646_0yw1r8/index.html
The figures there are really great!

The run on this branch (which happened to be the comparison) complained about a memory leak but I didn't get a similar complaint about the branch merged with `master` so I think we're good.  Probably unrelated to this PR but something to keep an eye out for.

**Comment by mkstratos** [comment]
Thanks for all your work on this @xylar! Apologies for the branch name, next time I'll be more "creative". I've made the change you requested, it works with the EVV test and produces the correct output.

**Review by xylar — APPROVED** [review]
Excellent!  Looks good to me!

**Comment by rljacob** [comment]
@jonbob can you start merging this?

**Comment by jonbob** [comment]
The new test worked for me as well, once I sourced the correct file. Also passes sanity testing.

Merged to next

**Comment by jonbob** [comment]
merged to master

**Pull request merged**